toolchain/check/context.cpp

```diff
@@ -445,95 +449,61 @@ auto Context::ImplicitAsForArgs(
 
   // If sizes mismatch, fail early.
   if (arg_refs.size() != param_refs.size()) {
-    CARBON_CHECK(diagnostic != nullptr) << "Should have validated first";
-    CARBON_DIAGNOSTIC(CallArgCountMismatch, Note,
+    CARBON_DIAGNOSTIC(CallArgCountMismatch, Error,
                       "Function cannot be used: Received {0} argument(s), but "
                       "require {1} argument(s).",
                       int, int);
-    diagnostic->Note(param_parse_node, CallArgCountMismatch, arg_refs.size(),
-                     param_refs.size());
+    CARBON_DIAGNOSTIC(InCallToFunction, Note,
+                      "Calling function declared here.");
+    emitter_
+        ->Build(call_parse_node, CallArgCountMismatch, arg_refs.size(),
+                param_refs.size())
+        .Note(param_parse_node, InCallToFunction)
+        .Emit();
     return false;
   }
 
   // Check type conversions per-element.
-  // TODO: arg_ir_id is passed so that implicit conversions can be inserted.
-  // It's currently not supported, but will be needed.
   for (auto [i, value_id, param_ref] : llvm::enumerate(arg_refs, param_refs)) {
-    auto as_type_id = semantics_ir_->GetNode(param_ref).type_id();
-    if (ImplicitAsImpl(value_id, as_type_id,
-                       diagnostic == nullptr ? &value_id : nullptr) ==
-        ImplicitAsKind::Incompatible) {
-      CARBON_CHECK(diagnostic != nullptr) << "Should have validated first";
-      CARBON_DIAGNOSTIC(CallArgTypeMismatch, Note,
-                        "Function cannot be used: Cannot implicitly convert "
-                        "argument {0} from `{1}` to `{2}`.",
-                        size_t, std::string, std::string);
-      diagnostic->Note(param_parse_node, CallArgTypeMismatch, i,
-                       semantics_ir_->StringifyType(
-                           semantics_ir_->GetNode(value_id).type_id()),
-                       semantics_ir_->StringifyType(as_type_id));
-      return false;
-    }
+    DiagnosticAnnotationScope annotate_diagnostics(
```

Comment:
Moved the scope outside the loop as suggested.

Regarding this scope versus passing information into the callee: I'm concerned that the complexity of passing around a note-printing mechanism would be substantial, as the volume of code that's reachable through parameter initialization (and other situations where we have context notes) grows -- we'd end up passing it to probably most or all of the code in `check` that can produce diagnostics, so having the scope change some state in the emitter seems better. Forwarding around an extra parameter is also likely to add up to a fair bit of overhead even if no diagnostic is emitted.

I'm expecting we'll want to change the implementation on the diagnostics side a bit (or perhaps a lot) in the future. For example, I think it would be a good idea for the diagnostics mechanism to hold a separate list of context callbacks rather than swapping out the diagnostic consumer. In Clang, we also only print the context note for the first diagnostic that's emitted within the scope, and we might want to consider making that change in the future as well. But I think the general model of (cheaply) registering some kind of callback, somewhere in the diagnostics infrastructure, is probably the right way to go.

Reply:
> I think it would be a good idea for the diagnostics mechanism to hold a separate list of context callbacks rather than swapping out the diagnostic consumer.

I see you had the same suggestion in a later comment. Done :-)